File: services/client/src/client/protocol.go
package client

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"io"

	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/safe_socket"
)

const TLV_BET_TYPE uint16 = 0x01
const TLV_END_TYPE uint16 = 0x02
const TLV_WINNER_TYPE uint16 = 0x03
const TLV_NEW_BATCH_TYPE uint16 = 0x04
const TLV_ACK_TYPE uint16 = 0x05
const TLV_NACK_TYPE uint16 = 0x06
const EXPECTED_FIELDS = 6
const TLV_HEADER_SIZE = 4

func serialize_tlv_message(msgType uint16, payload []byte) ([]byte, error) {
	message := new(bytes.Buffer)
	if err := binary.Write(message, binary.BigEndian, msgType); err != nil {
		return nil, err
	}
	if err := binary.Write(message, binary.BigEndian, uint16(len(payload))); err != nil {
		return nil, err
	}
	message.Write(payload)
	return message.Bytes(), nil
}

func serialize_end_message(agencyId string) ([]byte, error) {
	payload := []byte(agencyId)
	return serialize_tlv_message(TLV_END_TYPE, payload)
}

func serialize_bet(bet Bet) ([]byte, error) {
	fields := []string{
		bet.AgencyId,
		bet.FirstName,
		bet.LastName,
		bet.Id,
		bet.Birthdate,
		bet.BetNumber,
	}

	payload := new(bytes.Buffer)
	// Build body
	for i, field := range fields {
		// Type
		if err := binary.Write(payload, binary.BigEndian, uint16(i+1)); err != nil {
			return nil, err
		}
		// Size
		if err := binary.Write(payload, binary.BigEndian, uint16(len(field))); err != nil {
			return nil, err
		}
		// Value
		payload.WriteString(field)
	}

	return serialize_tlv_message(TLV_BET_TYPE, payload.Bytes())
}

func serialize_batch(bets []Bet) ([]byte, error) {
	payload := new(bytes.Buffer)
	for _, bet := range bets {
		serializedBet, err := serialize_bet(bet)
		if err != nil {
			return nil, err
		}
		_, err = payload.Write(serializedBet)
		if err != nil {
			return nil, err
		}
	}
	return serialize_tlv_message(TLV_NEW_BATCH_TYPE, payload.Bytes())
}

func deserialize_ack(socket io.Reader) error {
	header, err := safe_socket.RecvAll(socket, TLV_HEADER_SIZE)
	if err != nil {
		return err
	}

	tlvType := binary.BigEndian.Uint16(header[0:2])
	tlvSize := int(binary.BigEndian.Uint16(header[2:4]))

	if tlvSize > 0 {
		return fmt.Errorf("formato de mensaje incorrecto")
	}

	switch tlvType {
	case TLV_ACK_TYPE:
		return nil
	case TLV_NACK_TYPE:
		return fmt.Errorf("el servidor rechazó el batch")
	default:
		return fmt.Errorf("tipo de mensaje inesperado")
	}
}

func deserialize_winners(socket io.Reader) ([]Bet, error) {
	header, err := safe_socket.RecvAll(socket, TLV_HEADER_SIZE)
	if err != nil {
		return []Bet{}, err
	}

	tlvType := binary.BigEndian.Uint16(header[0:2])
	tlvSize := int(binary.BigEndian.Uint16(header[2:4]))

	switch tlvType {
	case TLV_WINNER_TYPE:
		tlvValue, err := safe_socket.RecvAll(socket, tlvSize)
		if err != nil {
			return []Bet{}, err
		}
		return extractBets(tlvValue)
	default:
		return []Bet{}, fmt.Errorf("tipo de mensaje inesperado")
	}
}

func extractBet(rawBet []byte) (Bet, error) {
	index := 0
	var elems []string

	for index < len(rawBet) {
		if index+TLV_HEADER_SIZE > len(rawBet) {
			return Bet{}, fmt.Errorf("formato de apuesta incorrecto")
		}
		length := int(binary.BigEndian.Uint16(rawBet[index+2 : index+4]))
		index += TLV_HEADER_SIZE
		if index+length > len(rawBet) {
			return Bet{}, fmt.Errorf("formato de apuesta incorrecto")
		}
		content := string(rawBet[index : index+length])
		elems = append(elems, content)
		index += length
	}

	if len(elems) != EXPECTED_FIELDS {
		return Bet{}, fmt.Errorf("formato de apuesta incorrecto")
	}

	return Bet{
		AgencyId:  elems[0],
		FirstName: elems[1],
		LastName:  elems[2],
		Id:        elems[3],
		Birthdate: elems[4],
		BetNumber: elems[5],
	}, nil
}

func extractBets(payload []byte) ([]Bet, error) {
	index := 0
	var bets []Bet

	for index < len(payload) {
		if index+TLV_HEADER_SIZE > len(payload) {
			return []Bet{}, fmt.Errorf("formato de apuesta incorrecto")
		}
		length := int(binary.BigEndian.Uint16(payload[index+2 : index+4]))
		index += TLV_HEADER_SIZE
		if index+length > len(payload) {
			return []Bet{}, fmt.Errorf("formato de apuesta incorrecto")
		}
		content := payload[index : index+length]
		bet, err := extractBet(content)
		if err != nil {
			return []Bet{}, err
		}
		bets = append(bets, bet)
		index += length
	}

	return bets, nil
}

File: services/server/src/server/protocol.py
import safe_socket
from lottery.bet import Bet

TLV_BET_TYPE = 0x01
TLV_END_TYPE = 0x02
TLV_WINNERS_TYPE = 0x03
EXPECTED_FIELDS = 6
TLV_HEADER_SIZE = 4

def serialize_tlv_message(msg_type, payload):
    header = msg_type.to_bytes(2, byteorder='big') + len(payload).to_bytes(2, byteorder='big')
    return header + bytes(payload)

def serialize_bet(bet):
    fields = [
        str(bet.agency_id),
        str(bet.first_name),
        str(bet.last_name),
        str(bet.document),
        str(bet.birthdate),
        str(bet.number),
    ]
    
    payload = bytearray()
    for field_index, field in enumerate(fields, start=1):
        field_bytes = field.encode("utf-8")
        payload.extend(field_index.to_bytes(2, byteorder='big'))
        payload.extend(len(field_bytes).to_bytes(2, byteorder='big'))
        payload.extend(field_bytes)

    return serialize_tlv_message(TLV_BET_TYPE, payload)


def serialize_winners(winners):
    payload = bytearray()
    for winner in winners:
        bet_bytes = serialize_bet(winner)
        # Ignore NEW_BET header
        payload.extend(bet_bytes[TLV_HEADER_SIZE:])
    return serialize_tlv_message(TLV_WINNERS_TYPE, payload)


def deserialize(socket):
    header = safe_socket.recv_all(socket, TLV_HEADER_SIZE)
    if not header:
        return None

    tlv_type = int.from_bytes(header[0:2], byteorder='big')
    tlv_size = int.from_bytes(header[2:4], byteorder='big')

    if tlv_type == TLV_BET_TYPE:
        tlv_value = safe_socket.recv_all(socket, tlv_size)
        return {
            "type": "NEW_BET",
            "data": extract_bet(tlv_value),
        }
    elif tlv_type == TLV_END_TYPE:
        tlv_value = safe_socket.recv_all(socket, tlv_size)
        agency_id = tlv_value.decode('utf-8')
        return {
            "type": "END_BETS",
            "data": agency_id,
        }
    elif tlv_type == TLV_WINNERS_TYPE:
        # manejar error
        pass
    else:
        raise ValueError(f"Tipo de mensaje no reconocido: {tlv_type}")


def extract_bet(bet):
    index = 0
    elems = []
    while index < len(bet):
        length = int.from_bytes(bet[index+2:index+4], byteorder='big')
        index += TLV_HEADER_SIZE
        content = bet[index:index + length].decode("utf-8")
        elems.append(content)
        index += length

    return Bet(
        agency_id=int(elems[0]),
        first_name=elems[1],
        last_name=elems[2],
        document=int(elems[3]),
        birthdate=elems[4],
        number=int(elems[5]),
    )


